### ISSUE TITLE
Support for keyfile

### ISSUE THREAD

**Issue by benruijl** [guns/go-luks-suspend#3]
I tried the package and it seems to be working! The only thing I couldn't get working yet is that the program first tries to read a keyfile before asking for the password.

I am using the following option:
````
cryptkey=/dev/sdb:12892:2048
````

The `encrypt` hook used on boot first tries this device. If it isn't there, it asks for the password.

I ran go-luks-suspend in Virtualbox, so it could also be that the issue lies there.

**Comment by guns** [comment]
That sounds easy to add!

Just so I'm clear though,

1) The keyfile for the root volume is stored on a removable block device (e.g. USB drive).

2) You boot with this drive, then remove it.

3) You suspend the system, wiping crypt keys in memory.

4) Before wake (or after) you reattach this device.

5) Cryptsetup unlocks the root volume with the keyfile from the device. *No password is necessary to unlock the keyfile*.

Is that right?

**Comment by benruijl** [comment]
That's right, with one addition: if the device is not plugged in for a given amount of time,  the password should be prompted. The password is configured as a secondary unlock option for my root volume.

It would be really cool if the password is always prompted, but as soon as the device is inserted, it takes precedence and unlocks the system. This may be impossible though.

**Comment by guns** [comment]
> It would be really cool if the password is always prompted, but as soon as the device is inserted, it takes precedence and unlocks the system. This may be impossible though.

That shouldn't be too hard. I don't know offhand how to receive hardware notifications from `udev`, but we can always poll in a loop as a last resort.

**Comment by benruijl** [comment]
I should perhaps make one thing a bit more clear: my 'keyfile' is actually just a certain position on a block device. It exists between partitions and is thus not really a file. I see you have added support for keyfiles now, which is great and will help a lot of people, but doesn't quite work my use case.

**Comment by guns** [comment]
> I should perhaps make one thing a bit more clear: my 'keyfile' is actually just a certain position on a block device.

Ah, you were quite clear, I am just tackling this issue in stages. :)

I am planning on replicating the functionality of the [encrypt][] hook, so this is the plan:

1) Add keyfile unlocking with offsets and sizes for non-root volumes. ✔

2) Add keyfile unlocking with offsets and sizes for the root volume with keyfiles that live directly on a block device.

3) Add keyfile unlocking with offsets and sizes for the root volume with keyfiles that live in a filesystem (requires mount, read, unmount).

4) Create a polling thread/event handler that reacts to block device insertions, attempts to unlock, and cancels the main unlock thread.

After this, I'd like to add support for the `sd-encrypt` hook.

I have uncommitted changes for `(2)`, but I probably won't be able to continue work on go-luks-suspend until after the weekend. I am enjoying this project, so don't lose hope!

[encrypt]: https://git.archlinux.org/svntogit/packages.git/tree/trunk/encrypt_hook?h=packages/cryptsetup

**Comment by guns** [comment]
Okay, you should be able to unlock the root device using a keyfile specified by:

```
cryptkey=device:offset:size
```

I decided against automatically unlocking on device hotplug. Instead, you press `CTRL-R` to retry unlocking with a keyfile. Alternately, waking the system with the block device already attached will unlock the root volume without a password prompt.

Unlocking with a keyfile specified by `cryptkey=device:filesystem:path` is still forthcoming.

**Comment by guns** [comment]
`cryptkey=device:filesystem:path` is now supported! Thanks for encouraging me to add keyfile support for the root volume. Please reopen this issue if you find any bugs.

**Comment by Risebell** [comment]
First of all: nice package and mostly working for me! But I'm not quite sure how using a cryptkey works.
If I understand correctly it should read the Kernel commandline and use the exact same setting as used for a regular boot, right? If so, then that doesn't work out for me.

My cryptkey line looks like this: `cryptkey=UUID=5D3C-1461:vfat:/filename`
I guess it's related to either using a UUID or to vfat as the filesystem, but I can't figure out how to troubleshoot this.

Thanks in advance!

**Comment by benruijl** [comment]
I believe that currently a key file is only supported if it is not in a
file system, but just at a certain position on disk.

**Comment by Risebell** [comment]
> cryptkey=device:filesystem:path is now supported! Thanks for encouraging me to add keyfile support for the root volume. Please reopen this issue if you find any bugs.

That was guns last message here, so I assume it should be supported..

**Comment by guns** [comment]
@benruijl Keyfile support should be implemented for both.

@Risebell I tested this feature with exactly that kind of `cryptkey` (USB drive specified with UUID=, fs with vfat), so your parameter style is not an issue.

Could you please run the binary in debug mode and share the redacted output?

```
# /usr/lib/go-luks-suspend/go-luks-suspend -debug
```

**Comment by Risebell** [comment]
```
2017/09/20 20:24:04 gathering cryptdevices
2017/09/20 20:24:04 Name:"system" IsRootDevice:true
2017/09/20 20:24:04 running pre-suspend scripts
2017/09/20 20:24:04 gathering filesystems with write barriers
2017/09/20 20:24:04 main.filesystem{mountpoint:"/", devno:0x18}
2017/09/20 20:24:04 main.filesystem{mountpoint:"/home", devno:0x2c}
2017/09/20 20:24:04 main.filesystem{mountpoint:"/btrfs_pool", devno:0x17}
2017/09/20 20:24:04 checking for suspend program in initramfs
2017/09/20 20:24:04 preparing initramfs chroot
2017/09/20 20:24:04 stopping selected system services
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active syslog.socket
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active systemd-journald.socket
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active systemd-journald-dev-log.socket
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active systemd-journald-audit.socket
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active systemd-journald.service
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active systemd-udevd-control.socket
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active systemd-udevd-kernel.socket
2017/09/20 20:24:04 exec: /usr/bin/systemctl --quiet is-active systemd-udevd.service
2017/09/20 20:24:04 exec: /usr/bin/systemctl stop systemd-journald.socket systemd-journald-dev-log.socket systemd-journald.service systemd-udevd-control.socket systemd-udevd-kernel.socket systemd-udevd.service
2017/09/20 20:24:04 stopped systemd-journald.socket, systemd-journald-dev-log.socket, systemd-journald.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-udevd.service
2017/09/20 20:24:04 flushing pending writes
2017/09/20 20:24:04 disabling write barriers on filesystems to avoid IO hangs
2017/09/20 20:24:04 calling suspend in initramfs chroot
2017/09/20 18:24:04 starting udevd from initramfs
2017/09/20 18:24:04 exec: /usr/lib/systemd/systemd-udevd --daemon --resolve-names=never
2017/09/20 18:24:04 loading cryptdevice names
2017/09/20 18:24:04 suspending cryptdevices
2017/09/20 18:24:04 suspending system
2017/09/20 18:24:04 exec: /usr/bin/cryptsetup luksSuspend system
2017/09/20 18:24:04 debug: skipping suspend to RAM
2017/09/20 18:24:04 resuming root cryptdevice

Press Escape to suspend to RAM, or Ctrl-R to rescan block devices for keyfiles, or Ctrl-T to start a debug shell.

Enter passphrase for system: 2017/09/20 18:24:04 exec: /usr/bin/cryptsetup --tries=1 luksResume system

Attempting to unlock system with keyfile...

Press Escape to suspend to RAM, or Ctrl-R to rescan block devices for keyfiles, or Ctrl-T to start a debug shell.

Enter passphrase for system: 2017/09/20 18:24:15 exec: /usr/bin/cryptsetup --tries=1 luksResume system

2017/09/20 18:24:35 stopping udevd within initramfs
2017/09/20 18:24:35 exec: /usr/bin/udevadm control --exit
2017/09/20 20:24:35 starting previously stopped system services
2017/09/20 20:24:35 exec: /usr/bin/systemctl start systemd-journald.socket systemd-journald-dev-log.socket systemd-journald.service systemd-udevd-control.socket systemd-udevd-kernel.socket systemd-udevd.service
2017/09/20 20:24:35 gathering keyfiles from /etc/crypttab
2017/09/20 20:24:35 goLuksSuspend.Keyfile{Path:"/filename", Device:"/dev/disk/by-uuid/5D3C-1461", FSType:"vfat", Offset:0, Size:0}
2017/09/20 20:24:35 resuming cryptdevices with keyfiles
2017/09/20 20:24:35 re-enabling write barriers on filesystems
2017/09/20 20:24:35 unmounting initramfs bind mounts
2017/09/20 20:24:35 running post-suspend scripts
```

That's the output I get. As far as I can see it seems to recognize the keyfile, but the output only mentions it after I typed in my passphrase.

**Comment by guns** [comment]
Thank you for posting the debug log. `go-luks-suspend` apparently sees your device, but also doesn't attempt to use it.

Could you please:

1) Start a debug shell at the prompt with `CTRL-T`.

2) Verify that `/dev/disk/by-uuid/5D3C-1461` is available.

3) Plug and unplug the device containing the keyfile.

4) Verify step (2) again.

5) Exit the shell with `CTRL-D`, and try to unlock with the keyfile.

The primary purpose of the above to determine if `/dev/disk/by-uuid/5D3C-1461` is available but still unused by `go-luks-suspend`.

Thank you.

**Comment by Risebell** [comment]
So I think I figured something out:
the real filename of my keyfile has a dot in it, I just didn't want to show that filename here for reasons. The above steps didn't work, as in the device was listed in `/dev/disk/by-uuid/`, before and after unplugging and plugging it back in, but after exiting the shell and hitting Ctrl-R it showed me it has unlocked the root filesystem and still asked for the passphrase.
I renamed the keyfile to "keyfile" and set the kernel cmd accordingly and now if I am at the password prompt and unplug, plug and then hit Ctrl-R it actually unlocks with the keyfile.
Note I have to unplug and plug, else it doesn't recognize it.

**Comment by guns** [comment]
> So it seems like the keyfile isn't recognized if the device stays in my laptop. Is that wanted behaviour?

No, this is definitely a bug. A strange one, since `go-luks-suspend` can see the device, but somehow decides against using it. My VM doesn't display this behavior, so I will try to reproduce this on actual hardware.

Thanks for the report.

**Comment by guns** [comment]
Unfortunately, I couldn't replicate this bug on my hardware or in a VM.

> Just checked, I have to plug it in as I am at the unlock screen.

This seems to imply that the copy of systemd-udevd started in the initramfs is not recognizing hotplug events until after the prompt is displayed. I've pushed a change to the [issue3](https://github.com/guns/go-luks-suspend/tree/issue3) branch that calls `udevadm trigger; udevadm settle` in the initramfs in hopes that your udev daemon will receive the hotplug events from the kernel and respond accordingly.

Could you please test this change?

> Interestingly, in the case the storage device stays in the laptop it tells that it's attempting to unlock with the keyfile and then still asks for the passphrase, opposing to the case if I unplug it completly where it says it can't find the keyfile.

The message, "Attempting to unlock %s with keyfile...", is displayed immediately after `CTRL-R` if the device can be `stat()`ed successfully. Strangely, the same `stat()` check fails moments later when we decide whether we should attempt a keyfile unlock!

In order to debug this, I've also pushed a change that dumps the results of each `stat()` call. Feel free to redact this info when you post the debug log like you did above.

Thanks!

**Comment by Risebell** [comment]
Umm.. Now I feel like my System is kinda strange anyway.
So I wanted to test the modified version today, but first tried it out again with the standard version (1.4.0) from aur; I didn't update anything in between.
I don't get this behaviour anymore and I think it's working as it should now.
I'm sorry I can't replicate the bug myself anymore...

But I guess I found another bug: if I am at the passphrase prompt and hit Ctrl-C it kills the prompt and I get access to my Xorg session, without a keyfile and without entering the passphrase. Should I open a new issue for that? Is it the same for you or is my System in general just kinda weird?

I'm sorry for any inconvenience...

**Comment by guns** [comment]
> I'm sorry I can't replicate the bug myself anymore...

If it happened once, it can happen again, so I'm glad that you reported it.

> if I am at the passphrase prompt and hit Ctrl-C it kills the prompt

Yes, I think exiting the whole program on SIGINT is surprising. I'll add a handler that just cancels the prompt.

> and I get access to my Xorg session, without a keyfile and without entering the passphrase

Screen locking is out-of-scope for this project; I recommend installing something like `slock` to lock your screen on suspend.

The systemd service file for `go-luks-suspend` uses `openvt` to simply launch the binary in a new virtual terminal:

```
ExecStart=/usr/bin/openvt -ws -- /usr/lib/go-luks-suspend/go-luks-suspend
```

This means that there is nothing stopping you from switching back to your original console using `CTRL-ALT-F{1..12}` after you wake from sleep.

You would, however, quickly find that every process that would like to read or write to your disk will hang indefinitely since the encrypted device has been locked.

Thanks for reminding me about `CTRL-C`! I'll fix that in `v1.4.1`.

**Comment by Risebell** [comment]
Ah, thanks for clarifying! I thought I shouldn't be able to get out of the prompt screen at all.
I'm glad I could help and I will report if I find any other bugs, I really appreciate the project!
